internal/pkg/service/stream/storage/level/local/volume/disksync/disksync.go

```diff
@@ -16,50 +15,77 @@ import (
 	"github.com/keboola/keboola-as-code/internal/pkg/utils/errors"
 )
 
-// Syncer writes data to the underlying writer and according to the Config it calls the chain.Flush() or chain.Sync().
-// Regarding waiting for sync, see Notifier and DoWithNotifier methods.
+// Syncer according to the Config calls the Chain.Flush() or Chain.Sync().
+// Regarding waiting for sync, see Notifier methods.
 type Syncer struct {
 	logger log.Logger
 	clock  clock.Clock
 	config Config
-	chain  chain
+	chain  Chain
 
-	ctx    context.Context
-	cancel context.CancelFunc
-	wg     *sync.WaitGroup
+	stopped chan struct{}
+	wg      *sync.WaitGroup
 
-	// writeOpsCount is updated via AddWriteOp method.
-	// It contains count of high-level writers, e.g., one table row = one write operation.
-	// This may not correspond to the number of calls of the low-level Write or WriteString methods.
-	writeOpsCount *atomic.Uint64
-	// lastSyncAt is updated on each sync start
-	lastSyncAt *atomic.Time
-	// bytesToSync are updated by the Write method, the Syncer is the first writer in the chain
-	bytesToSync *atomic.Uint64
+	statistics StatisticsProvider
+
+	// Snapshots of the metrics at the last sync
+
+	acceptedWritesSnapshot   *atomic.Uint64
+	uncompressedSizeSnapshot *atomic.Uint64
+	compressedSizeSnapshot   *atomic.Uint64
+	lastSyncAt               *atomic.Time
 
 	// syncLock ensures that only one sync runs at a time
 	syncLock *sync.Mutex
 
-	// notifierLock ensures that each high-level write operation receives the notifier that belongs to it, see DoWithNotifier.
+	// notifierLock protects the notifier field during a swap on sync
 	notifierLock *sync.RWMutex
 	notifier     *notify.Notifier
 
+	// opFn is operation triggered on sync
 	opFn func(ctx context.Context) error
 }
 
-// chain is a resource responsible for synchronizing of file writers.
-type chain interface {
-	io.Writer
-	io.StringWriter
+type SyncerFactory func(
+	ctx context.Context,
+	logger log.Logger,
+	clock clock.Clock,
+	config Config,
+	chain Chain,
+	statistics StatisticsProvider,
+) *Syncer
+
+// Chain is a resource that should be synchronized.
+type Chain interface {
 	// Flush data from memory to OS disk cache. Used if Config.Mode=ModeToCache.
```

Comment:
How this works? I dont understand how we are able to write into disk cache. I though it is abstraction and OS says whether it goes into cache or right into disk itself.

Reply:
Try to look at the resulting code, the diff is quite large.